python/lsst/daf/butler/core/config.py

```diff
@@ -217,10 +225,35 @@ def __initFromFile(self, path):
             To a persisted config file.
         """
         if path.endswith("yaml"):
-            self.__initFromYamlFile(path)
+            if path.startswith('s3://'):
```

Comment:
Changed to use ButlerURI class instead. I disagree with renaming the method part to `__initFromUri` for two reasons:

 * first, because they are files - just at different locations
 * second, `__initFromFile` compliments other instantiation methods such as initializing form another Config, or lets say from another stream. It calls the appropriate method that then actually inits the config. In the case of `initFromYamlFile` it determined that the file format is yaml and initiates from Yaml file. If support for pickle was added this would be the place to make that determination. The `initFromYamlFile` then opens the file as a stream and lets `initFromYaml` parse it to a Config. Because of this I changed the docs to better reflect the fact that path can be path or an uri. I have renamed `initFromS3File` to `initFromS3YamlFile` and have changed the method to call `initFromYaml` following how it was previsouly done and make that distinction explicit.

This brings about a different problem and that is that the boto3 HTML response object is a `StreamingObject` (implemented in boto3) but it itself does not seem to be a valid Python IO stream. I have added the minimum attributes missing (the `name` attribute) explicitly but this feels a bit hackish. On first glance none of the other important methods/properties/attributes are missing except for `seek` functionality that a proper stream would have. I don't know if I should be worried that `Loader` in `lsst.daf.butler.config` will ever be changed sufficiently for this to matter. If it does, we can always change back to reading the yaml directly in that method like it was previously done (but the code is less well structured then), or writing a better wrapper for `StreamingBody`.